### PR TITLE
If we re-fakeify a FakeTensor with the same ShapeEnv, preserve symbols

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113690
* #113080
* #111554
* #113666
* #113653
* __->__ #113651

Subsumes half of https://github.com/pytorch/pytorch/pull/113605

We support fakeifying an already fake tensor, which will give you a new fake tensor mirroring the same structure as the original fake tensor, which is what is needed by https://github.com/pytorch/pytorch/issues/113643 . However, when this refakeification happens, we will naively reallocate all new sizes for all of the fake tensor. This is the right thing to do if you are re-fakeifying on a fresh ShapeEnv (because you're reparametrizing the sizes or something), but if you have two fake tensor modes which are sharing a shape environment, you would actually rather just reuse the original sizes/strides/offset from the original fake tensor. This ends up being pretty simple. I recommend viewing with whitespace diff turned off.

There's some fuzz around jagged tensor handling; that code is probably not quite right, but I fixed it for this particular case in the most straightforward way.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

**Review by albanD — APPROVED** [review]
nit could have a test where the shape env don't match and ensure new symints are re-generated?

**Review by thiagocrepaldi — COMMENTED** [review]
https://github.com/pytorch/pytorch/pull/113605#issuecomment-1810640367

File: torch/_subclasses/meta_utils.py [hunk 1]
```diff
@@ -474,31 +483,43 @@ def empty_create(inner_t, inner_src):
...
+                            assert isinstance(ctx, dict)
+                            assert "ragged_size" in ctx
+                            assert isinstance(t._size[1], torch.SymInt)
+                            if orig_shape_env is shape_env:
+                                # It's already fake and the shape envs line up, reuse the old size
+                                # Do not assert singleton_int; it may already
+                                # be a variable
+                                ctx["ragged_size"] = t._size[1]
+                            else:
+                                assert t._size[1].node.singleton_int() is not None
```
Comment (eellison) [inline torch/_subclasses/meta_utils.py:511]: I guess for this to be in the generic mechanism, we would need to change the api of `__tensor_flatten__` ?

**Review by eellison — APPROVED** [review]

File: torch/_subclasses/meta_utils.py (hunk 1 above)
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:511]: Yeah, I feel there is something wrong with the code here, but not entirely sure what. I mentioned this to @soulitzer. At least part of the problem is that the ragged special case code doesn't correctly respect DimDynamic

**Review by bdhirsh — APPROVED** [review]

File: torch/_subclasses/meta_utils.py (hunk 1 above)
Comment (jbschlosser) [reply torch/_subclasses/meta_utils.py:511]: In #113279, I'm working on correctly respecting DimDynamic

**Pull request closed without merging**